File: Interpreter.java
import java.io.*;
import java.util.Scanner;

class Token
{
  public static String INTEGER = "INTEGER";
  public static String PLUS = "PLUS";
  public static String MINUS = "MINUS";
  public static String MULTIPLY = "MULTIPLY";
  public static String DIVIDE = "DIVIDE";
  public static String EOF = "EOF";

  String value,type;
  public Token(String value,String type){
    this.value = value;
    this.type = type;
  }

  public String getValue(){
    return this.value;
  }
  public String getType(){
    return this.type;
  }
}

class Interpreter{
  String text;
  Token currentToken;
  int pos;

  public Interpreter(String text){
    this.text = text.replaceAll("\\s+","");
    pos = 0;
  }

  public Token getNextToken()throws Exception{

    if(pos>text.length()-1){
      return new Token("",Token.EOF);
    }

    char currentChar = text.charAt(pos);
    String multi = ""+currentChar;
    if(Character.isDigit(currentChar)){
      Token token = new Token(String.valueOf(currentChar),Token.INTEGER);
      pos += 1;
      while (pos < text.length() && Character.isDigit(text.charAt(pos))){
        currentChar = text.charAt(pos);
        multi +=currentChar;
        token = new Token(multi,Token.INTEGER);
        if(pos < (text.length()-1))
          pos += 1;
        else
          break;
      }
      return token;
    }
    else if(currentChar == '+'){
      Token token = new Token(String.valueOf(currentChar),Token.PLUS);
      pos += 1;
      return token;
    }
    else if(currentChar == '-'){
      Token token = new Token(String.valueOf(currentChar),Token.MINUS);
      pos += 1;
      return token;
    }
    else if(currentChar == '*'){
      Token token = new Token(String.valueOf(currentChar),Token.MULTIPLY);
      pos += 1;
      return token;
    }
    else if(currentChar == '/'){
      Token token = new Token(String.valueOf(currentChar),Token.DIVIDE);
      pos += 1;
      return token;
    }
    else{
      error();
      return null;
    }
  }

  public void error()throws Exception{
    throw new Exception();
  }

  public void eat(String tokenType)throws Exception{
    if(currentToken.getType() == tokenType){
      currentToken = getNextToken();
    }
    else{
      error();
    }
  }

  public String term()throws Exception{
    Token token = currentToken;
    eat(Token.INTEGER);
    return token.getValue();
  }

  public int expr()throws Exception{
    currentToken = getNextToken();
    int result = Integer.parseInt(term());

    if(currentToken.getType() == Token.PLUS ||
    currentToken.getType() == Token.MINUS){
    while(currentToken.getType() == Token.PLUS ||
    currentToken.getType() == Token.MINUS){
      if(currentToken.getType() == Token.PLUS){
        eat(Token.PLUS);
        result = result + Integer.parseInt(term());
      }
      if(currentToken.getType() == Token.MINUS){
        eat(Token.MINUS);
        result = result - Integer.parseInt(term());
      }
    }
  }

  if(currentToken.getType() == Token.MULTIPLY ||
  currentToken.getType() == Token.DIVIDE){
  while(currentToken.getType() == Token.MULTIPLY ||
  currentToken.getType() == Token.DIVIDE){
    if(currentToken.getType() == Token.MULTIPLY){
      eat(Token.MULTIPLY);
      result = result * Integer.parseInt(term());
    }
    if(currentToken.getType() == Token.DIVIDE){
      eat(Token.DIVIDE);
      result = result / Integer.parseInt(term());
    }
  }
}
    return result;
  }

  public static void main(String [] args) throws Exception
  {
    while(true){
    Scanner reader = new Scanner(System.in);
    System.out.println("Enter an arithmetic expression: ");
    String exp = reader.nextLine();
    Interpreter interpreter = new Interpreter(exp);
    int res = interpreter.expr();
    System.out.println(res);
  }
  }
}

File: Interpreter.py
# Token types
#
# EOF (end-of-file) token is used to indicate that
# there is no more input left for lexical analysis
INTEGER, PLUS, MINUS, MULTIPLY, DIVIDE, EOF = 'INTEGER', 'PLUS', 'MINUS', 'MULTIPLY', 'DIVIDE', 'EOF'


class Token(object):
    def __init__(self, type, value):
        # token type: INTEGER, PLUS, MINUS, MULTIPLY, DIVIDE or EOF
        self.type = type
        # token value: 0, 1, 2. 3, 4, 5, 6, 7, 8, 9, '+','-','*','/' or None
        self.value = value

    def __str__(self):
        """String representation of the class instance.

        Examples:
            Token(INTEGER, 3)
            Token(PLUS '+')
        """
        return 'Token({type}, {value})'.format(
            type=self.type,
            value=repr(self.value)
        )

    def __repr__(self):
        return self.__str__()


class Interpreter(object):
    def __init__(self, text):
        # client string input, e.g. "3+5"
        self.text = "".join(text.split())

        # self.pos is an index into self.text
        self.pos = 0
        # current token instance
        self.current_token = None

    def error(self):
        raise Exception('Error parsing input')

    def get_next_token(self):
        """Lexical analyzer (also known as scanner or tokenizer)

        This method is responsible for breaking a sentence
        apart into tokens. One token at a time.
        """
        text = self.text

        # is self.pos index past the end of the self.text ?
        # if so, then return EOF token because there is no more
        # input left to convert into tokens
        if self.pos > len(text) - 1:
            return Token(EOF, None)

        # get a character at the position self.pos and decide
        # what token to create based on the single character
        current_char = text[self.pos]

        # if the character is a digit then convert it to
        # integer, create an INTEGER token, increment self.pos
        # index to point to the next character after the digit,
        # and return the INTEGER token
        if current_char.isdigit():
            token = Token(INTEGER, int(current_char))
            self.pos += 1
            while(self.pos<len(text) and text[self.pos].isdigit()):
                current_char = current_char + text[self.pos]
                token = Token(INTEGER, int(current_char))
                self.pos += 1
            return token

        if current_char == '+':
            token = Token(PLUS, current_char)
            self.pos += 1
            return token

        if current_char == '-':
            token = Token(MINUS, current_char)
            self.pos += 1
            return token
        if current_char == '*':
            token = Token(MULTIPLY, current_char)
            self.pos += 1
            return token
        if current_char == '/':
            token = Token(DIVIDE, current_char)
            self.pos += 1
            return token
        self.error()

    def eat(self, token_type):
        # compare the current token type with the passed token
        # type and if they match then "eat" the current token
        # and assign the next token to the self.current_token,
        # otherwise raise an exception.
        if self.current_token.type == token_type:
            self.current_token = self.get_next_token()
        else:
            self.error()



    def expr(self):
        """expr -> ARITHMETIC EXPRESSION"""
        # set current token to the first token taken from the input
        self.current_token = self.get_next_token()
        left = self.current_token
        self.eat(INTEGER)

        while self.pos<len(self.text):
        # we expect the current token to be a '+','-','*','/' token
            op = self.current_token
            if op.type == PLUS:
                self.eat(PLUS)
            elif op.type == MINUS:
                self.eat(MINUS)
            elif op.type == MULTIPLY:
                self.eat(MULTIPLY)
            else:
                self.eat(DIVIDE)
            # we expect the current token to be an integer
            right = self.current_token
            self.eat(INTEGER)


            # at this point INTEGER PLUS INTEGER sequence of tokens
            # has been successfully found and the method can just
            # return the result of adding two integers, thus
            # effectively interpreting client input
            if op.type == PLUS:
                result = left.value + right.value
            elif op.type == MINUS:
                result = left.value - right.value
            elif op.type == MULTIPLY:
                result = left.value * right.value
            else:
                result = left.value / right.value
            left = Token(INTEGER, int(result))

        return result


    def term(self):
        token= self.current_token
        self.eat(INTEGER)
        return token.value

    def expr2(self):
        self.current_token = self.get_next_token()
        result = self.term()
        while self.current_token.type in (PLUS, MINUS, MULTIPLY, DIVIDE):
            token = self.current_token
            if token.type == PLUS:
                self.eat(PLUS)
                result = result + self.term()
            elif token.type == MINUS:
                self.eat(MINUS)
                result = result - self.term()
            elif token.type == MULTIPLY:
                self.eat(MULTIPLY)
                result = result * self.term()
            else:
                self.eat(DIVIDE)
                result = result / self.term()

        return result


def main():
    while True:
        try:
            # To run under Python3 replace 'raw_input' call
            # with 'input'
            text = raw_input('calc> ')
        except EOFError:
            break
        if not text:
            continue
        interpreter = Interpreter(text)
        result = interpreter.expr2()
        print(result)


if __name__ == '__main__':
    main()
